### DISCUSSION TITLE
Passing Data To Future Steps in case of temporary file paths and logger

### DISCUSSION THREAD

**Discussion by Sax388** [spring-projects/spring-batch#4577] [Q&A] [answered]
I'd like to get some advice on https://docs.spring.io/spring-batch/reference/common-patterns.html#passingDataToFutureSteps.
- To produce a report I'd need to prepare several files (report file(s) and log file) and pack them into a ZIP file afterwards.
- For the preparation I create a temporary folder and put the files there to zip it later and then copy it to the output directory
  - So the path of the temporary directory needs to be shared among the steps (write and zip step)
  - And the Logger object needs to be shared so we can initialize and close it in accordance with the job

Is it a good idea to use the `ExecutionContextPromotionListener` for this use case?
- In the example from the docs the variable is put into the context for every write. This seems a bit expensive for a non-changing path.

**Comment by fmbenhassine** [comment]
Yes I think so. Once the temporary folder is created, its path can be promoted to the job execution context so it can be shared with subsequent steps.

**Reply by Sax388** [reply]
But why add it for every write. I don't really get the example here. Could it also be a job parameter that is created when the job is triggered by e.g. a Quartz scheduler?

**Reply by fmbenhassine** [reply, marked as answer]
While the example in the reference documentation updates the key on every write (like updating a counter for instance), this is not necessary. In your case, if the preparation step is a simple tasklet that creates the temporary folder, then the path of the folder can be put in the job execution context only once.

Using a job parameter should work as well if the temporary folder is created before launching the job (to be able to pass the path of the folder as a job parameter).
